src/geff/writer_helper.py

```diff
@@ -0,0 +1,104 @@
+import abc
+from typing import Any, Sequence
+
+import numpy as np
+import zarr
+from numpy.typing import ArrayLike
+
+
+class BaseWriterHelper(abc.ABC):
+    """
+    Base class for writing geff attributes to a zarr group.
+
+    Important:
+        This function does not create the `ids` attribute.
+
+    Args:
+        group: The zarr group to write the attribute to.
+        position_attr: The name of the position attribute.
+
+    Raises:
+        ValueError: If the group is not a 'nodes' or 'edges' group.
+    """
+
+    def __init__(self, group: zarr.Group, position_attr: str | None):
+        if group.name not in ["/nodes", "/edges"]:
+            raise ValueError(f"Group must be a 'nodes' or 'edges' group. Found {group.name}")
+
+        self._group = group
+        self._position_attr = position_attr
+
+    @abc.abstractmethod
+    def _data_dict_iterator(self) -> Sequence[tuple[Any, dict[str, Any]]]:
+        """
+        Methods to iterate over the data dictionary to be written.
+        Must match the index ordering
+        """
+
+    @abc.abstractmethod
+    def _attr_names(self) -> list[str]:
+        """
+        Methods to iterate over the attribute names to be written.
+        """
+
+    def write_attrs(self) -> None:
```

Comment:
So this being generalizable seems to rely on the ability of the in memory representation being able to provide the Sequence data structure of tuples of (id, attr_dict). I just want to double check that there aren't any edge cases that we might be missing where that would be inconvenient. Although if I think of working with a dataframe in memory this would still work.

Reply:
An iterator would still work as long as you could iterate multiple times through it.